### DISCUSSION TITLE
Add option to disable mipmaps on raster layers

### DISCUSSION THREAD

**Discussion by kircher1** [maplibre/maplibre-gl-js#1828] [Ideas]
Mipmaps can improve quality by reducing texture aliasing, but at some cost, like 33% additional GPU memory per texture plus computation overhead on the client that is required to generate the mip levels.

Currently mipmaps are enabled by default, but could an option be added to the styling to disable it? 

```
    "mipmap": {
      "type": "boolean",
      "default": true,
      "doc": "Enables or disables mipmaps.",
     ....
      }
```

If so, one point to clarify would be whether this is a property of the `paint_raster` or `source_rastser`... I could see an argument for both, but because mipmapping requires an explicit call to GL for generating miplevels on the texture, this seems possibly more appropriate as a property on the source.

**Comment by HarelM** [comment]
Can you link to an example that shows the current usage/problem?

**Reply by zhangyiatmicrosoft** [reply]
The root problem is the TEXTURE_MIN_FILTER, example:
https://jsbin.com/hofoqoz/edit?html,output

The last one is an ordinary img element as baseline while the other three are using different filters.
All of them are the same on standard monitor setting (device pixel ratio of 1, 1.5 or 2).

If user has an uncommon setting, or use ctrl+wheel to adjust the browser scale, when 1< DPR < 1.5, you can see the quality difference. 
MapLibre hardcoded LINEAR_MIPMAP_NEAREST, which can be less crisp than LINEAR.

The request here is to add an option, something like "raster-min-resampling" to allow each layer to specify it. Thanks

**Reply by kircher1** [reply]
Thanks @zhangyiatmicrosoft for sharing that as it underlines the visual component to this conversation.

We should let user specify the minification filter, just as can be done with the magnification filter via `raster-resampling` paint property-- The minification filter should be specified in such a way that it allows the raster layer to avoid generating mips if the filter does not use them, and thus avoids the perf overhead of generating mips when they aren't actually used.

**Reply by HarelM** [reply]
Unfortunately, this is beyond my knowledge. I would recommend posting a link to this discussion in slack to get more input.

**Comment by kircher1** [comment]
This snippet in the Texture class is what ultimately kicks off mipmap generation: https://github.com/maplibre/maplibre-gl-js/blob/224a8c86c8bb3b9abe2a95193b840e710eeba3b6/src/render/texture.ts#L75

The Texture class already has a constructor parameter for enabling/disabling the mipmap, but for raster sources, their Textures are always created with `useMipmap= true`, and there's no way to disable it.

What I'd like is some way to configure `useMipmap`... Maybe styling is the right avenue for that, but the setting for raster-resampling is a similar idea, controlling how texture sampling works: https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#paint-raster-raster-resampling

**Comment by kircher1** [comment]
I created a PR that uses the config system (rather than styling) to control whether or not mipmaps are generated for raster textures: https://github.com/maplibre/maplibre-gl-js/pull/1853
